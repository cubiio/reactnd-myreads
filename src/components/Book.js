import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Icon from '../icons/icons';

const BooksGridLI = styled.li`
  padding: 10px 15px;
  text-align: left;
`;

const StyledBook = styled.div`
  width: 140px;
`;

const StyledBookTop = styled.div`
  display: flex;
  align-items: flex-end;
  position: relative;
  height: 200px;
`;

const StyledBookCover = styled.div`
  /* background-image contained in style tag in the JSX */
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  background: #eee;
  width: 128px;
  height: 193px;
`;

const StyledBookShelfChanger = styled.div`
  position: absolute;
  right: 0;
  bottom: -10px;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background: #60ac5d;
  background-image: url(${Icon.arrowDropDownIcon});
  background-repeat: no-repeat;
  background-position: center;
  background-size: 20px;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
`;

const StyledBookShelfChangerSelect = styled.select`
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
`;

const StyledBookInfo = styled.div`
  font-size: 0.8em;
`;

const StyledBookTitle = StyledBookInfo.extend`
  margin-top: 10px;
`;

const StyledBookAuthors = StyledBookInfo.extend`
  color: #999;
`;

const Book = (props) => {
  const { authors, shelf, title } = props.book;
  const { bookImgUrl, onUpdateBook } = props;

  const handleMoveBook = (e) => {
    const selectedBook = props.book;
    const selectedShelf = e.target.value;
    onUpdateBook(selectedBook, selectedShelf);
  };

  const formattedAuthorStr = (writers) => {
    if (writers === undefined) {
      return writers;
    }
    return writers.length >= 2 ? writers.join(', ') : writers;
  };

  return (
    <BooksGridLI>
      <StyledBook>
        <StyledBookTop>
          <StyledBookCover
            style={{
              backgroundImage: `url(${bookImgUrl})`,
            }}
          />
          <StyledBookShelfChanger>
            <StyledBookShelfChangerSelect
              value={shelf || 'none'}
              onChange={handleMoveBook}
            >
              <option disabled>Move to...</option>
              <option value="currentlyReading">Currently Reading</option>
              <option value="wantToRead">Want to Read</option>
              <option value="read">Read</option>
              <option value="none">None</option>
            </StyledBookShelfChangerSelect>
          </StyledBookShelfChanger>
        </StyledBookTop>
        <StyledBookTitle>{title}</StyledBookTitle>
        <StyledBookAuthors>{formattedAuthorStr(authors)}</StyledBookAuthors>
      </StyledBook>
    </BooksGridLI>
  );
};

Book.propTypes = {
  book: PropTypes.object,
  bookImgUrl: PropTypes.string,
  onUpdateBook: PropTypes.func.isRequired,
};

Book.defaultProps = {
  book: {},
  bookImgUrl: '',
};

export default Book;
