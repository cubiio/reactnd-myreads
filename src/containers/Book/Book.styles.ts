import styled from 'styled-components';
import Icon from '../../icons/icons';

export const BooksGridLI = styled.li`
  padding: 10px 15px;
  text-align: left;
`;

export const StyledBook = styled.div`
  width: 140px;
`;

export const StyledBookTop = styled.div`
  display: flex;
  align-items: flex-end;
  position: relative;
  height: 200px;
`;

export const StyledBookCover = styled.div`
  /* background-image contained in style tag in the JSX */
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  background: #eee;
  width: 128px;
  height: 193px;
`;

export const StyledBookShelfChanger = styled.div`
  position: absolute;
  right: 0;
  bottom: -10px;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background: #60ac5d;
  background-image: url(${Icon.arrowDropDownIcon});
  background-repeat: no-repeat;
  background-position: center;
  background-size: 20px;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
`;

export const StyledBookShelfChangerSelect = styled.select`
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
`;

export const StyledBookInfo = styled.div`
  font-size: 0.8em;
`;

export const StyledBookTitle = StyledBookInfo.extend`
  margin-top: 10px;
`;

export const StyledBookAuthors = StyledBookInfo.extend`
  color: #999;
`;
