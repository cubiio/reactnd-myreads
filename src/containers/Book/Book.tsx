import * as React from 'react';
import * as Styles from './Book.styles';
import {
  BookComponentProps,
  MoveBookEvent,
} from '../../interfaces/bookComponent';

const Book = (props: BookComponentProps) => {
  const { authors, shelf, title } = props.book;
  const { bookImgUrl, onUpdateBook } = props;

  const handleMoveBook = (e: MoveBookEvent) => {
    const selectedBook = props.book;
    const selectedShelf = e.target.value;
    onUpdateBook(selectedBook, selectedShelf);
  };

  const formattedAuthorStr = (writers: string[]) => {
    if (writers === undefined) {
      return writers;
    }
    return writers.length >= 2 ? writers.join(', ') : writers;
  };

  const disabled: boolean = true;
  return (
    <Styles.BooksGridLI>
      <Styles.StyledBook>
        <Styles.StyledBookTop>
          <Styles.StyledBookCover
            style={{
              backgroundImage: `url(${bookImgUrl})`,
            }}
          />
          <Styles.StyledBookShelfChanger>
            <Styles.StyledBookShelfChangerSelect
              value={shelf || 'none'}
              onChange={handleMoveBook}
            >
              <option disabled={disabled}>Move to...</option>
              <option value="currentlyReading">Currently Reading</option>
              <option value="wantToRead">Want to Read</option>
              <option value="read">Read</option>
              <option value="none">None</option>
            </Styles.StyledBookShelfChangerSelect>
          </Styles.StyledBookShelfChanger>
        </Styles.StyledBookTop>
        <Styles.StyledBookTitle>{title}</Styles.StyledBookTitle>
        <Styles.StyledBookAuthors>
          {formattedAuthorStr(authors)}
        </Styles.StyledBookAuthors>
      </Styles.StyledBook>
    </Styles.BooksGridLI>
  );
};

export default Book;
