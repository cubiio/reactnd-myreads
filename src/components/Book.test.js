/* global describe, it, expect */

import React from 'react';
import renderer from 'react-test-renderer';
import Book from './Book';

describe('Book', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Book />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
