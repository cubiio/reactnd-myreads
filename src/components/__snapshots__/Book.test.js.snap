// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book renders correctly 1`] = `
<li
  className="sc-bdVaJa kjQISI"
>
  <div
    className="sc-bwzfXH cTLMOu"
  >
    <div
      className="sc-htpNat fvAskg"
    >
      <div
        className="sc-bxivhb jVTYef"
        style={
          Object {
            "backgroundImage": "url()",
          }
        }
      />
      <div
        className="sc-ifAKCX bJLWbW"
      >
        <select
          className="sc-EHOje RUfBm"
          onChange={[Function]}
          value="wantToRead"
        >
          <option
            disabled={true}
          >
            Move to...
          </option>
          <option
            value="currentlyReading"
          >
            Currently Reading
          </option>
          <option
            value="wantToRead"
          >
            Want to Read
          </option>
          <option
            value="read"
          >
            Read
          </option>
          <option
            value="none"
          >
            None
          </option>
        </select>
      </div>
    </div>
    <div
      className="sc-gzVnrw jhGaQF"
    >
      A Book
    </div>
    <div
      className="sc-htoDjs cnGTWw"
    >
      Author A
    </div>
  </div>
</li>
`;
